File: BehavioralAnalysis_python/EMG/EMGIndex.py
# -*- coding: utf-8 -*-

import numpy as np

class EMG_Index:
    
    def EMG_Indexing(xds, muscle_groups):
        
        #%% Define the muscles
        
        muscle_names = []
        
        if muscle_groups == 'Flex':
            muscle_names.append('FCR')
            muscle_names.append('FCU')
            
        if muscle_groups == 'Exten':
            muscle_names.append('ECR')
            muscle_names.append('ECU')
            
        if muscle_groups == 'Both':
            muscle_names.append('FCR')
            muscle_names.append('FCU')
            muscle_names.append('ECR')
            muscle_names.append('ECU')
            
        if muscle_groups == 'Uln_Dev':
            muscle_names.append('FCU')
            muscle_names.append('ECU')
            
        if muscle_groups == 'Rad_Dev':
            muscle_names.append('FCR')
            muscle_names.append('ECR')
        
        if muscle_groups == 'Grasp':
            muscle_names.append('FDP')
            muscle_names.append('FDS')
        
        if muscle_groups == 'Custom':
            muscle_names.append('FDS1')
        
        # If muscle_names is still blank
        if muscle_names == []:
            muscle_names.append(muscle_groups)
            
        # Find the indices of the muscles of interest
        muscle_idx = []
        cc = 0
        for ii in range(len(muscle_names)):
            muscle_idx.append([idx for idx, s in enumerate(xds.EMG_names) if muscle_names[ii] in s])
            # Find how many muscles there are
            cc = cc + len(muscle_idx[ii])
            
        # Concatenate the indices
        M = []
        for ii in range(len(muscle_idx)):
            for jj in range(len(muscle_idx[ii])):
                M.append(muscle_idx[ii][jj])
            
        if muscle_groups == 'All':
            M = list(range(len(xds.EMG_names)))
            
            
        return M
    
    
    

File: BehavioralAnalysis_matlab/EMG/EMG_Index.m
function [M] = EMG_Index(xds, muscle_groups)

%% File Description:

% This function finds the indices of relevant EMG in an XDS file based on
% the muscle groups you are interested in looking at
%
% -- Inputs --
% xds: the XDS file in question
% muscle_groups: 'Flex', 'Ext', Uln_Dev', 'Rad_Dev', 'Both', 'Grasp', 'Custom', or 'All'

%% If there is no EMG

if ~xds.has_EMG
    disp('No EMG in this file');
    M = [];
    return
end

%% Define the muscles 

muscle_names = strings;

if strcmp(muscle_groups, 'Flex')
    muscle_names(1) = 'FCR';
    muscle_names(2) = 'FCU';
end

if strcmp(muscle_groups, 'Ext')
    muscle_names(1) = 'ECR';
    muscle_names(2) = 'ECU';
end

if strcmp(muscle_groups, 'Uln_Dev')
    muscle_names(1) = 'FCU';
    muscle_names(2) = 'ECU';
end

if strcmp(muscle_groups, 'Rad_Dev')
    muscle_names(1) = 'FCR';
    muscle_names(2) = 'ECR';
end

if strcmp(muscle_groups, 'Both')
    muscle_names(1) = 'FCR';
    muscle_names(2) = 'FCU';
    muscle_names(3) = 'ECR';
    muscle_names(4) = 'ECU';
end

if strcmp(muscle_groups, 'Grasp')
    muscle_names(1) = 'FDP';
    muscle_names(2) = 'FDS';
end

if strcmp(muscle_groups, 'Pinch')
    muscle_names(1) = '1DI';
    muscle_names(2) = 'APB';
    muscle_names(3) = 'FPB';
end

if strcmp(muscle_groups, 'Custom')
    muscle_names(1) = 'FCR';
    muscle_names(2) = 'FDP';
    muscle_names(3) = 'FDS';
    muscle_names(4) = 'EDC';
end

% If muscle_names is still blank
if (muscle_names == "")
    muscle_names(1) = muscle_groups;
end

% Find the indices of the muscles of interest
muscle_idx = struct([]);
cc = 0;
for ii = 1:length(muscle_names)
    muscle_idx{ii,1} = find(contains(xds.EMG_names, muscle_names(ii)));
    % Find how many muscles there are
    cc = cc + length(muscle_idx{ii,1});
end

% Concatenate the indices
M = zeros(cc,1);
cc = 1;
for ii = 1:length(muscle_idx)
    for jj = 1:length(muscle_idx{ii})
        M(cc) = muscle_idx{ii,1}(jj);
        cc = cc + 1;
    end
end

if strcmp(muscle_groups, 'All')
    M = 1:length(xds.EMG_names);
    M = M';
end

